source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc

```diff
@@ -73,89 +196,165 @@ void Filter::onReadWorker() {
     } else {
       throw EnvoyException("failed to read proxy protocol");
     }
+  }
+}
 
-    // Error check the source and destination fields. Most errors are caught by the address
-    // parsing above, but a malformed IPv6 address may combine with a malformed port and parse as
-    // an IPv6 address when parsing for an IPv4 address. Remote address refers to the source
-    // address.
-    const auto remote_version = remote_address->ip()->version();
-    const auto local_version = local_address->ip()->version();
-    if (remote_version != protocol_version || local_version != protocol_version) {
-      throw EnvoyException("failed to read proxy protocol");
+bool Filter::parseExtensions(int fd) {
+  while (proxyProtocolHeader_.value().extensions_length) {
+    // buf_ is no longer in use so we re-use it to read/discard
+    int bytes_avail;
+    auto& os_syscalls = Api::OsSysCallsSingleton::get();
+    if (os_syscalls.ioctl(fd, FIONREAD, &bytes_avail) < 0) {
+      throw EnvoyException("failed to read proxy protocol (no bytes avail)");
     }
-    // Check that both addresses are valid unicast addresses, as required for TCP
-    if (!remote_address->ip()->isUnicastAddress() || !local_address->ip()->isUnicastAddress()) {
-      throw EnvoyException("failed to read proxy protocol");
+    if (bytes_avail == 0) {
+      return false;
     }
-
-    // Only set the local address if it really changed, and mark it as address being restored.
-    if (*local_address != *socket.localAddress()) {
-      socket.setLocalAddress(local_address, true);
+    bytes_avail = std::min(size_t(bytes_avail), sizeof(buf_));
+    bytes_avail = std::min(size_t(bytes_avail), proxyProtocolHeader_.value().extensions_length);
+    ssize_t nread = os_syscalls.recv(fd, buf_, bytes_avail, 0);
+    if (nread == -1 && errno == EAGAIN) {
+      return false;
+    } else if (nread != bytes_avail) {
+      throw EnvoyException("failed to read proxy protocol extension");
     }
-    socket.setRemoteAddress(remote_address);
+    proxyProtocolHeader_.value().extensions_length -= nread;
   }
-
-  // Release the file event so that we do not interfere with the connection read events.
-  file_event_.reset();
-  cb_->continueFilterChain(true);
+  return true;
 }
 
-bool Filter::readLine(int fd, std::string& s) {
+bool Filter::readProxyHeader(int fd) {
   while (buf_off_ < MAX_PROXY_PROTO_LEN) {
-    ssize_t nread = recv(fd, buf_ + buf_off_, MAX_PROXY_PROTO_LEN - buf_off_, MSG_PEEK);
+    int bytes_avail;
+    auto& os_syscalls = Api::OsSysCallsSingleton::get();
+
+    if (os_syscalls.ioctl(fd, FIONREAD, &bytes_avail) < 0) {
+      throw EnvoyException("failed to read proxy protocol (no bytes avail)");
+    }
+
+    if (bytes_avail == 0) {
+      return false;
+    }
+
+    bytes_avail = std::min(size_t(bytes_avail), MAX_PROXY_PROTO_LEN - buf_off_);
+
+    ssize_t nread = os_syscalls.recv(fd, buf_ + buf_off_, bytes_avail, MSG_PEEK);
 
     if (nread == -1 && errno == EAGAIN) {
```

Comment:
Now that you are calling ioctl first, should this ever actually return EAGAIN?

Reply:
um, i suppose its unlikely.
EAGAIN technically means any reason that it would block, e.g. insufficient buffer space, e.g. insufficient data received. But in practise in Linux it just means not enough data received for readlen, i'm not sure for other POSIX os if they follow the spec or the subset of it.

would you like me to make this fatal to the connection? it seems better as-is.